### PR TITLE
Yearn V2 share price trigger

### PR DESCRIPTION
**Pull request by mds1**
Adds a Yearn V2 trigger based on `pricePerShare`. This is similar to Compound's `exchangeRateStored` trigger from #3, but with an important difference: In Yearn V2 vaults, the pricePerShare decreases immediately after a harvest, and typically ramps up over the next six hours. Therefore we cannot simply check that the pricePerShare increases. Instead, we consider the vault triggered if the pricePerShare drops by more than 50% from it's previous value. This is conservative, but previous Yearn bugs resulted in pricePerShare drops of 0.5% – 10% ([source](https://github.com/yearn/yearn-security/tree/master/disclosures)), and these were only temporary drops with users able to be made whole. Therefore this trigger requires a large 50% drop to minimize false positives

(Yearn V1 vaults should never have a decrease in share price. [This tweet](https://twitter.com/real_Mills/status/1369794597803671554) and the replies show the difference well)

### CODE REVIEW

File: contracts/YearnV2SharePrice.sol [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * @notice Defines a trigger that is toggled if the price per share for the V2 yVault decreases between consecutive
+ * checks. Under normal operation, this value should only increase
+ */
+contract YearnV2SharePrice is ITrigger {
+  uint256 internal constant WAD = 10**18;
+
+  /// @notice Vault this trigger is for
+  IYVaultV2 public immutable market;
+
+  /// @notice Last read exchangeRateStored
```
Comment (apbendi) [inline contracts/YearnV2SharePrice.sol]: Comment is out of date? Should read `Last read pricePerShare`

File: contracts/YearnV2SharePrice.sol [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  uint256 public lastPricePerShare;
+
+  /// @dev In Yearn V2 vaults, the pricePerShare decreases immediately after a harvest, and typically ramps up over the
+  /// next six hours. Therefore we cannot simply check that the pricePerShare increases. Instead, we consider the vault
+  /// triggered if the pricePerShare drops by more than 50% from it's previous value. This is conservative, but
+  /// previous Yearn bugs resulted in pricePerShare drops of 0.5% – 10%, and were only temporary drops with users able
+  /// to be made whole. Therefore this trigger requires a large 50% drop to minimize false positives
+  uint256 public constant tolerance = 5e17; // 50%, represented on a scale where 1e18 = 100%
+
+  constructor(
```
Comment (apbendi) [inline contracts/YearnV2SharePrice.sol:31]: Consider adding natspec comments to the constructor, or at least a dev notice for the `_market` parameter.

File: contracts/YearnV2SharePrice.sol [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    // Save current exchange rate (immutables can't be read at construction, so we don't use `market` directly)
+    lastPricePerShare = IYVaultV2(_market).pricePerShare();
+  }
+
+  /**
+   * @dev Checks the yVault pricePerShare
+   */
+  function checkTriggerCondition() internal override returns (bool) {
+    // Read this blocks exchange rate
```
Comment (apbendi) [inline contracts/YearnV2SharePrice.sol]: Comments in this method also appear to be out of date, saying "exchange rate" when it should say "price per share," unless they're interchangeable. Also, in this particular line, I don't think "this blocks" is accurate either, as it's not constant in a given block.

File: scripts/deploy-yearn.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import { ethers, network } from 'hardhat';
+import axios from 'axios';
+import chalk from 'chalk';
+import readline from 'readline';
+
+const { BigNumber } = ethers;
+const { formatUnits, parseUnits } = ethers.utils;
+
+// Helper method to fetch JSON
+const fetch = (url: string) => axios.get(url).then((res) => res);
```
Comment (apbendi) [inline scripts/deploy-yearn.ts:11]: All of the helper methods seem to be duplicated across both deploy scripts. Does it make sense to pull them into a `deploy-utils.ts` file and import them? Might be helpful for future triggers as well.

**Review by apbendi — COMMENTED** [review]
Core functionality all looks good 👍. Left a handful of comments for your review.

File: contracts/YearnV2SharePrice.sol (hunk 1 above)
Comment (mds1) [reply contracts/YearnV2SharePrice.sol]: Fixed in 84899af1f94449f906b97263bbe6b72c9a00412c

File: contracts/YearnV2SharePrice.sol (hunk 2 above)
Comment (mds1) [reply contracts/YearnV2SharePrice.sol:31]: Fixed in 84899af1f94449f906b97263bbe6b72c9a00412c

File: contracts/YearnV2SharePrice.sol (hunk 3 above)
Comment (mds1) [reply contracts/YearnV2SharePrice.sol]: Fixed in 84899af1f94449f906b97263bbe6b72c9a00412c

File: scripts/deploy-yearn.ts (hunk 4 above)
Comment (mds1) [reply scripts/deploy-yearn.ts:11]: Agreed we should do that! But for now these are just quick and dirty deploy scripts. Ideally we'd have one deploy script would take contract name(s) as an input and read constructor arguments from a dedicated file (or something similar/scalable), that way it's easy to deploy one trigger for multiple markets, etc. But I think we should do that in a separate PR

Comment (apbendi) [reply scripts/deploy-yearn.ts:11]: Sounds good— we can make an issue and tackle it later.

**Review by apbendi — APPROVED** [review]

**Pull request merged**